src/Clapi/NamespaceTracker.hs

```diff
@@ -47,69 +63,54 @@ maybeNamespace :: (Name -> a) -> Path -> Maybe a
         when (isNamespace path) (modify $ \x -> Map.insert (namespace path) i x)
-    handle (RoutableMessage (Just Owner) (MsgDelete path)) =
+    handle (UMsgDelete path) =
         when (isNamespace path) (modify (Map.delete $ namespace path))
-    handle _ = return ()
 
--- FIXME: not filtering anything out so refactor accordingly
-registerSubscriptions ::
-    forall m i u. (Monad m, Ord i, Ord u) => AddrWithUser i u -> [RoutableMessage] -> StateT (Registered (AddrWithUser i u)) m [RoutableMessage]
-registerSubscriptions i rms = filterM processMsg rms
-  where
-    processMsg :: RoutableMessage -> StateT (Registered (AddrWithUser i u)) m Bool
-    processMsg (RoutableMessage Nothing (MsgAssignType path _)) =
-        modify (Mos.insert i path) >> return True
-    processMsg _ = return True
+registerSubscription ::
+    forall m i. (Monad m, Ord i) => i -> OwnerUpdateMessage -> StateT (Registered i) m ()
+registerSubscription i m = case m of
+    (Left (UMsgAssignType path _)) -> modify (Mos.insert i path)
+    _ -> return mempty
 
 fanOutBundle ::
-    (Monad m, Ord i, Ord u, Show i, Show u) =>
-    AddrWithUser i u ->
-    [RoutableMessage] ->
-    StateT (Owners (AddrWithUser i u), Registered (AddrWithUser i u)) (NsProtocol m i u) ()
-fanOutBundle awu rms = do
+    (Monad m, Ord i, Show i) =>
+    i ->
+    Maybe [UMsgError] ->
+    [OwnerUpdateMessage] ->
+    [UMsgError] ->
+    [OwnerUpdateMessage] ->
+    StateT (Owners i, Registered i) (NsProtocol m i) ()
+fanOutBundle i mOwnerErrs gets valErrs updates = do
     (po, ps) <- get
-    let allRecipients = Set.toList $ foldl Set.union (Set.singleton awu) [ownerAwus po, Map.keysSet ps]
-    let bundles = map (msgsFor rms po ps) allRecipients
-    let sendables = map (uncurry ServerData) $ filter (not . null . snd) $ zip allRecipients bundles
+    let allRecipients = Set.toList $ foldl Set.union (Set.singleton i) [owners po, Map.keysSet ps]
+    let bundles = map (bundleFor po ps) allRecipients
+    let sendables = map (uncurry ServerData) $ filter (nonEmptyBundle . snd) $ zip allRecipients bundles
     lift $ mapM_ sendRev sendables
   where
-    msgsFor rms po ps awu' = map rmMsg $ filter (includeMsg po ps awu') rms
-    includeMsg po ps awu' rm = maybe
-        (awu == awu')
-        (\o -> let p = msgPath' $ rmMsg rm in case o of
-            Owner -> isSubscriber ps awu' p
-            Client -> isOwner po awu' p)
-        (rmRoute rm)
-    isOwner po awu' p = Map.lookup (namespace p) po == Just awu'
-    isSubscriber ps awu' p = p `elem` Map.findWithDefault mempty awu' ps
-    ownerAwus po = Set.fromList $ Map.elems po
+    bundleFor po ps i' = UpdateBundle
+        (unicast valErrs i' ++ broadcastErrs ps i')
+        (unicast gets i' ++ updatesFor po ps i')
+    unicast ms i' = if i == i' then ms else []
+    broadcastSubs :: (Ord i, UMsg msg) => Registered i -> [msg] -> i -> [msg]
+    broadcastSubs ps ms i' = filter (\m -> uMsgPath m `elem` Map.findWithDefault mempty i' ps) ms
```

Comment:
I found the predicate in this filter being named `isSubscriber` easier to understand. I don't mind if the definition of `broadcastSubs` is `filter (isSubscriber blah) ms`, because that reads quite nicely.